### PR TITLE
feat: add PosNegLocalStorageManager as generic solution for managing …

### PR DESCRIPTION
**Pull request by wwalser**
Check this one first 🎉👍
The other two PRs that make up this end-to-end change are:
- sdk-react https://github.com/sajari/sdk-react/pull/599 
- search-widgets https://github.com/sajari/search-widgets/pull/159

### CODE REVIEW

**Comment by wwalser** [comment]
This this one first 🎉👍
The other two PRs that make up this end-to-end change are:
- sdk-react https://github.com/sajari/sdk-react/pull/599 
- search-widgets https://github.com/sajari/search-widgets/pull/159

File: src/index.ts [hunk 1]
```diff
@@ -1001,3 +1001,65 @@ export class Values extends EventEmitter {
...
+ * PosNegLocalStorageManager is a static utility class for manipulating Sajari's localstorage
+ * based management of PosNeg tokens. Typical use case is storing tokens for later consumption
+ * as users move through an ecommerce purcahse funnel.
+ */
+export class PosNegLocalStorageManager {
+  currentTokens: Record<string | number, TokenState>;
+  client: Client;
+  constructor(client: Client) {
+    const storageContent = localStorage.getItem(POS_NEG_STORAGE_KEY);
+    this.currentTokens = storageContent ? JSON.parse(storageContent) : {};
```
Comment (zlatanpham) [inline src/index.ts:1022]: Is that worth handling exception if by any chance `storageContent` is an invalid string?

File: src/index.ts [hunk 2]
```diff
@@ -1001,3 +1001,65 @@ export class Values extends EventEmitter {
...
+};
+
+export const POS_NEG_STORAGE_KEY = "sajari_tokens";
+
+/**
+ * PosNegLocalStorageManager is a static utility class for manipulating Sajari's localstorage
+ * based management of PosNeg tokens. Typical use case is storing tokens for later consumption
+ * as users move through an ecommerce purcahse funnel.
+ */
+export class PosNegLocalStorageManager {
```
Comment (zlatanpham) [inline src/index.ts:1017]: Do you think we need tests for this class?

File: src/index.ts [hunk 3]
```diff
@@ -1001,3 +1001,65 @@ export class Values extends EventEmitter {
...
+  token: PosNegToken;
+  clickSubmitted: boolean;
+};
+
+export const POS_NEG_STORAGE_KEY = "sajari_tokens";
+
+/**
+ * PosNegLocalStorageManager is a static utility class for manipulating Sajari's localstorage
+ * based management of PosNeg tokens. Typical use case is storing tokens for later consumption
+ * as users move through an ecommerce purcahse funnel.
```
Comment (tuanddd) [inline src/index.ts:1015]: typo purchase

File: src/index.ts [hunk 4]
```diff
@@ -1001,3 +1001,65 @@ export class Values extends EventEmitter {
...
+  get(fieldValue: string | number) {
+    return this.currentTokens[fieldValue];
+  }
+  sendPosEvent(
+    fieldValue: string | number,
+    identifier: string,
+    weight: number
+  ) {
+    const tokenState = this.get(fieldValue);
+    if (tokenState.clickSubmitted !== false) {
```
Comment (tuanddd) [inline src/index.ts:1041]: I think `tokenState` could be undefined?

**Review by tuanddd — COMMENTED** [review]
Noticing this class is called `POS_NEG_STORAGE_KEY` but it has only `sendPosEvent`, should we add another equivalent version for the negative token?

File: src/index.ts (hunk 2 above)
Comment (wwalser) [reply src/index.ts:1017]: Yeah, that's fair.

**Comment by wwalser** [comment]
> Noticing this class is called `POS_NEG_STORAGE_KEY` but it has only `sendPosEvent`, should we add another equivalent version for the negative token?

I've yet to see an actual use-case for neg tokens. I'm happy to leave using those as an exercise for the implementor or for the customer to specifically ask for this feature. 

Basically, I'm skeptical of their utility. If someone would benefit from them, they will let us know. Which will help me to learn more about their utility.

File: src/index.ts (hunk 1 above)
Comment (wwalser) [reply src/index.ts:1022]: Yup, fair.

File: src/index.ts (hunk 4 above)
Comment (wwalser) [reply src/index.ts:1041]: Shocked that TS didn't catch that. Thanks!

**Comment by tuanddd** [comment]
LGTM, +1!

**Comment by wwalser** [comment]
All comments resolved and tests written. Landing.

**Pull request merged**